### PR TITLE
Handle DicomFileLike

### PR DESCRIPTION
**Pull request by RobinFrcd**
Fixes https://github.com/ImagingDataCommons/highdicom/issues/222

Not sure about the expected behavior of `_check_file_format` as `fp` wasn't even used.

### CODE REVIEW

**Comment by CPBridge** [comment]
My guess is that the intention was that a DICOM file handle that is open (but not read into memory) could be passed as the argument. We would want to continue to support that case. Probably this means that the case of a `DICOMBytesIO` would be a further special case in the `if`...`else` chain. I agree that it would be a useful addition but it should not replace the existing case of an open file handle, and the logic for the two is somewhat different.

Just my initial thoughts, need to take a better look

**Comment by RobinFrcd** [comment]
> Out of interest, what is your goal with using the ImageFileReader here? The typical use case is one where you do not wish to read the entire file into memory because it may be very large. But in your case the object is already in memory as it is a `BytesIO`. Are you trying to avoid the cost of decompressing the full set of frames?

My use case here is that I download a DICOM file from a server and don't want to write it to disk. This file is 25MB on disk, 160MB on RAM when read by pydicom and 730MB if I load pixel_data ! So I want to keep the bytes on RAM and only read the frames I need with `highdicom` !

**Comment by RobinFrcd** [comment]
I've restored the previous case for `DicomFileLike`, all the tests are still passing :+1:

**Comment by RobinFrcd** [comment]
Hi,
Any news on this one ? I've been using my fork for a week now, working as expected ! 👍

File: src/highdicom/io.py
Comment (hackermd) [inline src/highdicom/io.py:367]: Let's only read the File Meta Information and avoid reading the entire file object (untested):

```python
        if self._filename is None:
            def stop_when(tag: pydicom.tag.BaseTag, vr: str, length: int) -> bool:
                if tag.group > 2:
                    return True
                return False
            file_meta = dcmread(fp, stop_when=stop_when).file_meta
            fp.seek(0)
        else:
            file_meta = read_file_meta_info(str(fp.name))
```

**Review by hackermd — CHANGES_REQUESTED** [review]
Thanks for the PR @RobinFrcd. I have made a few suggestions, which should fix an existing bug, ensure checks are run as intended, and improve performance.

I have not tested the suggestions, so please take them with a grain of salt.

File: src/highdicom/io.py
Comment (RobinFrcd) [reply src/highdicom/io.py:247]: I don't get this edit. Do you want to remove all the type checks on `filename` ?

File: src/highdicom/io.py
Comment (RobinFrcd) [reply src/highdicom/io.py]: Implemented in https://github.com/ImagingDataCommons/highdicom/pull/223/commits/d2eb66126facab94215377378783113506d770d5

File: tests/test_io.py
Comment (CPBridge) [inline tests/test_io.py:176]: We definitely need a test for the case where the input to the `ImageFileReader` is a `DicomFileLike` object that represents an opened file handle (and *not* a `DicomBytesIO`). I know this isn't really what you signed up for @RobinFrcd but I'd be more comfortable doing it in your branch so that we know everything is working before we merge. Therefore could you please accept this suggestion to add this test at the end of the file:

```suggestion
            np.testing.assert_array_equal(frame, pixel_array)
            
    def test_read_single_frame_ct_image_dicom_file_like_opened(self):
        # Test reading frames from an opened DicomFileLike file
        filename = self._test_dir.joinpath("ct_image.dcm")
        dcm = DicomFileLike(filename.open("rb"))

        dataset = dcmread(filename)
        pixel_array = dataset.pixel_array
        with ImageFileReader(dcm) as reader:
            assert reader.number_of_frames == 1
            frame = reader.read_frame(0)
            assert isinstance(frame, np.ndarray)
            assert frame.ndim == 2
            assert frame.dtype == np.int16
            assert frame.shape == (
                reader.metadata.Rows,
                reader.metadata.Columns,
            )
            np.testing.assert_array_equal(frame, pixel_array)
```

We should not merge until both new tests are passing

File: src/highdicom/io.py
Comment (CPBridge) [reply src/highdicom/io.py:257]: I'm also confused by this...

There are three cases that we need to consider:
- The plain str/`pathlib.Path` case
- The `DicomBytesIO` case (addressed in this PR)
- The `DicomFileLike` case where the `DicomFileLike` object is wrapped around an open file on disk. This case is currently broken on master and not fixed by this branch.

I was able to get all three cases to work with the following as my full `__init__` method:

```python
    def __init__(self, filename: Union[str, Path, DicomFileLike, DicomBytesIO]):
        """
        Parameters
        ----------
        filename: Union[str, pathlib.Path, pydicom.filebase.DicomfileLike]
            DICOM Part10 file containing a dataset of an image SOP Instance

        """
        if isinstance(filename, DicomFileLike):
            fp = filename
            if isinstance(filename, DicomBytesIO):
                self._fp = filename
                self._filename = None
            else:
                self._fp = fp
                self._filename = Path(fp.name)
        elif isinstance(filename, (str, Path)):
            self._filename = Path(filename)
            self._fp = None
        else:
            raise TypeError(
                'Argument "filename" must either an open DICOM file object or '
                'the path to a DICOM file stored on disk.'
            )
        self._metadata = None
```

However to get this to work I had to get rid of the tests on `is_little_endian` and `is_implicit_vr`. I don't think these are necessary (why would these characteristics not be read correctly from the file) but in the spirit of Chesterton's fence I'd rather understand from @hackermd why they were there in the first place before ripping them out.

**Review by CPBridge — CHANGES_REQUESTED** [review]
Thanks again @RobinFrcd for your work here. I have made a couple of minor suggestions that should be very quick to fix. I have also (in a suggestion) added a test for the third, unaddressed case of a `DicomFileLike` object wrapping an open part10 file that has not yet been read.

As I mention in one of the comments below, I just want confirmation from @hackermd about why the checks on `is_little_endian` and `is_implicit_vr` are necessary and whether it's ok to remove them

**Comment by CPBridge** [comment]
There are some small linter errors to deal with: https://github.com/ImagingDataCommons/highdicom/actions/runs/4689483856/jobs/8311389950?pr=223

**Review by CPBridge — APPROVED** [review]
Looks good to me, thanks @RobinFrcd for your quick work on this!

I approve but would prefer to wait for @hackermd  to have the final say as the original author of the `ImageFileReader`

File: src/highdicom/io.py
Comment (hackermd) [reply src/highdicom/io.py]: My point is that `DicomBytesIO` is a subtype of `DicomFileLike` and we only need to differentiate between the two types for setting `self._filename`. The other checks and setting of `self._fp` can be performed for both. I intended the checks to stay in place.

File: src/highdicom/io.py
Comment (hackermd) [inline src/highdicom/io.py:244]: Not needed because `DicomBytesIO` is a subtype of `DicomFileLike`.

```suggestion
    def __init__(self, filename: Union[str, Path, DicomFileLike]):
```

File: src/highdicom/io.py
Comment (hackermd) [inline src/highdicom/io.py:259]:
```suggestion
            self._fp = fp
            if isinstance(filename, DicomBytesIO):
                self._filename = None
            else:
                self._filename = Path(fp.name)
```

I suggest re-introducing the checks.

File: src/highdicom/io.py
Comment (hackermd) [inline src/highdicom/io.py:329]:
```suggestion
            fp: DicomFileLike
```

**Review by hackermd — CHANGES_REQUESTED** [review]

File: src/highdicom/io.py
Comment (CPBridge) [reply src/highdicom/io.py]: Understood, but as far as I can tell these checks never worked in the first place (and we missed this because we didn't have a test case for them). The checks did this:

1. Get values of `is_little_endian` and `is_implicit_vr` by reading the file meta using the `_check_file_format` method (which in turn uses the `pydicom.filereader.read_partial` method).
2. Get the same values by directly accessing (Python) attributes of the same name directly on the `fp` object, which will be a `DicomFileLike` object.
 
I'm currently failing to understand two points:
- Why would (2.) above ever work, since the `DicomFileLike` class does not have `is_little_endian` and `is_implicit_vr` attributes? I have played around with this and it always fails with an `AttributeError`
- If we imagine (2.) above did somehow work, under what circumstances would the results be different from the values obtained by (1.) above? In other words, why is check necessary?

It seems I am missing something big so I am unable to advise further except for just removing these checks, which as far as I can tell never actually worked and are unnecessary.

Comment (CPBridge) [reply src/highdicom/io.py]: Making some progress on understanding the confusion. What I said above is incorrect, `DicomFileLike` objects *do* have an `is_implicit_vr` attribute. Further, it does have an [`is_little_endian` property](https://github.com/pydicom/pydicom/blob/f7fdf0f12b235a315f850d58ab4e335a35f336b5/pydicom/filebase.py#L140), but this fails to return a value unless the [`is_little_endian` setter](https://github.com/pydicom/pydicom/blob/f7fdf0f12b235a315f850d58ab4e335a35f336b5/pydicom/filebase.py#L144) has previously explicitly been called on the object. This may actually be a bug in pydicom. This is what is causing my tests to fail whenever I tried it.

To demonstrate with the various ways of constructing a `DicomFileLike`

```python
from pydicom.filebase import DicomFile, DicomFileLike, DicomBytesIO
from pydicom.data import get_testdata_file

ct_file = get_testdata_file("ct_small.dcm")
print(ct_file)

# Using DicomFile function
dcm = DicomFile(ct_file, mode='rb')
assert isinstance(dcm, DicomFileLike)
print(dcm.is_implicit_VR)  # works
print(dcm.is_little_endian)  # fails

# Using DicomFileLike constructor on an open file
dcm = DicomFileLike(open(ct_file, mode='rb'))
assert isinstance(dcm, DicomFileLike)
print(dcm.is_implicit_VR) # works
# print(dcm.is_little_endian) # fails

# Using DicomBytesIO
dcm = DicomBytesIO(open(ct_file, "rb").read())
assert isinstance(dcm, DicomFileLike)
print(dcm.is_implicit_VR) # works
# print(dcm.is_little_endian) # fails
```

So in order to make this work as expected, the user would have to explicitly set the `is_little_endian` attribute before passing the object to the `ImageFileReader` class. Is this your intent @hackermd ? I would guess not...

Comment (hackermd) [reply src/highdicom/io.py]: Thanks for taking the deep-dive @CPBridge. Not sure whether it is a pydicom bug or feature, but I agree with you that the implementation of the `is_little_endian` logic is problematic. I would be fine with omitting the checks.

**Review by hackermd — APPROVED** [review]

**Comment by CPBridge** [comment]
Ok I'll merge. Thanks @RobinFrcd , we'll aim to have this in an upcoming patch release, probably after we fix #225

**Pull request merged**
